File: 20. Valid Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        
        for i in s:
            if i == '(' or i == '[' or i == '{':
                stack += [i]
            else:
                if (i == ')' and (len(stack) == 0 or stack[len(stack)-1] != '(')) or (i == ']' and (len(stack) == 0 or stack[len(stack)-1] != '[')) or (i == '}' and (len(stack) == 0 or stack[len(stack)-1] != '{')):
                    return False
                else:
                    stack.pop()
        return True if len(stack) == 0 else False

File: 20. Valid Parentheses.cpp
#include <stack>
class Solution {
public:
    bool isValid(string s) {
        stack <int> pstack;
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{')
                pstack.push(s[i]);
            else
            {
                if ((s[i] == ')' && (pstack.empty() || pstack.top() != '(')) || (s[i] == ']' && (pstack.empty() || pstack.top() != '[')) || (s[i] == '}' && (pstack.empty() || pstack.top() != '{')))
                    return false;
                else
                    pstack.pop();
            }
        }
        return pstack.empty() ? true : false;
    }
};
